File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
class Solution {
    public int solution(String s) {
        // n = 1000 --> O(N^2) 도 ㄱㅊ (완전탐색 수행)
        // 시뮬레이션 문제--> 문제에서 제시한 과정을 코드로 옮김 
        
        int answer = s.length() ; 
        
        // 1개 단위(step)부터 압축 단위를 늘려가며 확인 (절반 나누기까지)
        for(int step = 1 ; step < s.length() / 2 + 1 ; step++){
            String compressed = "" ; 
            String prev = s.substring(0, step) ; // 앞에서부터 step만큼의 문자열 추출 
            int cnt = 1 ; 
            
            
            for(int j = step; j < s.length() ; j+= step){
                // 단위 크기만큼 뒤의 문자열 잘라서 
                String sub ="" ; 
                for(int k = j ; k < j+step ; k++){
                    if(k < s.length())
                        sub+=s.charAt(k) ; 
                }
                // 이전 문자열과 같은지 확인 
                if(prev.equals(sub))  // 같으면
                    cnt++ ;  // 개수 증가 
                else{ // 다르면 
                    // 압축 진행 
                    compressed += (cnt >= 2)? cnt+prev: prev ; 
                    // 초기화 
                    sub = "" ; 
                    for(int k = j ; k < j+step ;k++){
                        if(k < s.length())
                            sub+= s.charAt(k) ; 
                    }
                    prev = sub ; 
                    cnt=1 ; 
                }
            }
            // 남아있는 문자열 처리 
            compressed += (cnt >= 2)? cnt+prev : prev; 
            answer = Math.min(answer, compressed.length()) ; 
        }
        return answer ; 
    }
}

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.py
def solution(S):
    # [0] 시뮬레이션 준비
    N = len(S)

    # debug = 0
    # print(S, N)

    # [1] 시뮬레이션 실행
    answer = 1000  # 최소 문자열 길이

    # 0) 자를 단위 정하고
    for k in range(1, N + 1):
        cnt = 1
        prev = ''
        new_S = ''
        # debug = 1
        # print(prev)
        # print(round(N / k))
        for i in range(N // k+2):
            # 1) 문자 추출
            cur = S[i * k:(i + 1) * k]
            # debug = 2
            # print(cur)

            # 2) 이전 문자 압축
            # print('bef',new_S)
            if prev == cur:
                cnt += 1
                # debug = 3
                # print(f"cnt={cnt}, prev={prev}, cur={cur}")

            else:
                if cnt > 1:
                    new_S += (str(cnt) + prev)
                elif cnt == 1:
                    new_S += prev
                cnt = 1
                prev = cur
                # debug = 4
                # print(new_S)

            # debug = 5
            # print('aft',new_S)

        # 4) min 갱신
        answer = min(answer, len(new_S))
        # debug = 6
        # print(answer, new_S, len(new_S))

    return answer
